### PR TITLE
Migrate datatree mapping.py

### PR DESCRIPTION
**Pull request by owenlittlejohns**
This PR continues the overall work of [migrating DataTree into xarray](https://github.com/pydata/xarray/issues/8572).

`datatree_mapping.py` is the renamed version of `mapping.py` from the `datatree` repository.

- [x] Closes migration step for `mapping.py` #8572
- [ ] ~~Tests added~~
- [ ] User visible changes (including notable bug fixes) are documented in `whats-new.rst`
- [ ] ~~New functions/methods are listed in `api.rst`~~

### CODE REVIEW

File: xarray/core/datatree_mapping.py [hunk 1]
```diff
@@ -4,10 +4,9 @@
 import sys
```
Comment (owenlittlejohns) [inline xarray/core/datatree_mapping.py:4]: @TomNicholas - there are no significant changes to the way `map_over_subtree` works in this PR. As discussed today, I've opened the PR, and am happy to make changes based on conversation here.

Comment (TomNicholas) [reply xarray/core/datatree_mapping.py:4]: Okay looking at this again I can't see an obvious way to refactor it. I'm mostly just concerned that it actually makes sense to you @owenlittlejohns ?

If so then I suggest we spend the time thinking about https://github.com/pydata/xarray/issues/8949 instead

Comment (owenlittlejohns) [reply xarray/core/datatree_mapping.py:4]: @TomNicholas - yeah. @flamingbear and I have both spent a fair bit of time trying to grok this module. I don't claim that I've got a perfect understanding, but I feel fairly good on the whole. (The conversation yesterday that led to #8949 helped a bit, too)

Comment (TomNicholas) [reply xarray/core/datatree_mapping.py:4]: Okay great. So I guess we merge this??

FYI if you want to chat about any of this over zoom outside of the weekly meeting I would be happy to.

Comment (owenlittlejohns) [reply xarray/core/datatree_mapping.py:4]: I think we're good to merge!

**Comment by TomNicholas** [comment]
Thank you!

**Pull request merged**